[Navigation] Navigation.navigate() should trigger the 'beforeunload' event synchronously

cdumez opened the pull request:
#### 4c2147c4b2aa67d561ab0b50feeefb8dc677cd35
<pre>
[Navigation] Navigation.navigate() should trigger the &apos;beforeunload&apos; event synchronously
<a href="https://bugs.webkit.org/show_bug.cgi?id=282593">https://bugs.webkit.org/show_bug.cgi?id=282593</a>
<a href="https://rdar.apple.com/139628818">rdar://139628818</a>

Reviewed by Rob Buis.

Navigation.navigate() should trigger the &apos;beforeunload&apos; event synchronously and quite
a few WPT tests were relying on this.

Fixing this enabled some more tests which exposed some pre-existing bugs that I&apos;m
fixing in the same PR so it doesn&apos;t look like regressions.

In particular, we dispatch events at NavigationHistoryEntry objects but the class
doesn&apos;t subclass ActiveDOMObject to keep the JS wrapper alive. We were also failing
to grab the JSLock in one place, which was causing assertions.

* LayoutTests/TestExpectations:
Unskip some tests that are now passing.

* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-handler-throws-expected.txt:
The output looks a bit worse but I don&apos;t think this is a regression. Rather, this looks like a timing change because
the navigation now happens sooner. I will investigate separately how to make this test fully pass.

* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/back-beforeunload-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/forward-beforeunload-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/navigate-beforeunload-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/navigate-rejection-order-beforeunload-unserializablestate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/navigate-rejection-order-invalidurl-beforeunload-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/reload-beforeunload-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/reload-rejection-order-beforeunload-unserializablestate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/traverseTo-beforeunload-expected.txt:
Rebaseline tests that are now passing.

* Source/WebCore/loader/FrameLoadRequest.h:
(WebCore::FrameLoadRequest::isFromNavigationAPI const):
(WebCore::FrameLoadRequest::setIsFromNavigationAPI):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::stopLoading):
(WebCore::FrameLoader::loadURL):
(WebCore::FrameLoader::loadWithDocumentLoader):
(WebCore::FrameLoader::continueLoadAfterNavigationPolicy):
* Source/WebCore/loader/FrameLoader.h:
* Source/WebCore/loader/NavigationAction.h:
(WebCore::NavigationAction::isFromNavigationAPI const):
(WebCore::NavigationAction::setIsFromNavigationAPI):
* Source/WebCore/loader/NavigationScheduler.cpp:
(WebCore::ScheduledHistoryNavigationByKey::findBackForwardItemByKey):
Make the navigation policy decision synchronous when the navigation is triggered via the navigation API.
This is the only way the navigation can proceed synchronously and fire the beforeunload event right away.

* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::stop):
When calling `window.stop()`, abort the pending navigation API navigation as tests are expecting this.

* Source/WebCore/page/NavigateEvent.cpp:
(WebCore::NavigateEvent::NavigateEvent):
(WebCore::NavigateEvent::create):
* Source/WebCore/page/NavigateEvent.h:
* Source/WebCore/page/NavigateEvent.idl:
Make sure the grab the JSLock before initializing m_info. Without this, I was seeing assertions
in the layout tests.

* Source/WebCore/page/Navigation.cpp:
* Source/WebCore/page/Navigation.h:
* Source/WebCore/page/NavigationHistoryEntry.cpp:
(WebCore::NavigationHistoryEntry::NavigationHistoryEntry):
(WebCore::NavigationHistoryEntry::create):
* Source/WebCore/page/NavigationHistoryEntry.h:
* Source/WebCore/page/NavigationHistoryEntry.idl:
Make NavigationHistoryEntry an ActiveDOMObject and keep its JS wrapper alive as long as the
NavigationHistoryEntry is held by the navigation object. This is important since the navigation
object may dispatch events at the NavigationHistoryEntry. Without this change, some of the tests
were asserting that the JS wrapper was missing during the event dispatching.

* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp:
(WebKit::WebFrameLoaderClient::dispatchDecidePolicyForNavigationAction):
Obey the policyDecisionMode being PolicyDecisionMode::Synchronous and use the DecidePolicyForNavigationActionSync
synchronous IPC when the navigation is triggered by the Navigation API so that the navigation can
start synchronously and yet still obey the policy decision from the client.

* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::dispatchDidFailProvisionalLoad):
Some of the tests abort the navigation during the provisional stage. This was hitting
an assertion in `WebLocalFrameLoaderClient::dispatchDidReachLayoutMilestone()`:
`ASSERT(!m_frame-&gt;isMainFrame() || webPage-&gt;corePage()-&gt;settings().suppressesIncrementalRendering() || m_didCompletePageTransition);`
The reason was the m_didCompletePageTransition wasn&apos;t reset to true and provisional load failure.
To address this issue, I now make sure to call `completePageTransitionIfNeeded()` in `dispatchDidFailProvisionalLoad()`.

Canonical link: <a href="https://commits.webkit.org/286919@main">https://commits.webkit.org/286919@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/21daa7d59e2da30209a8eb4eee2381c587ec2d4e

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/77485 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/56520 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/30401 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/82094 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/28767 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/79602 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/65669 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/4817 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/60730 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/18719 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/80552 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/50696 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/66510 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/41008 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/48097 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/24006 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/27090 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/69207 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/24344 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/83474 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/4865 "Built successfully") | [❌ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/3314 "Found 2 new test failures: ipc/create-connection-and-send-async.html webrtc/vp9-profile2.html (failure)") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/68957 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/5021 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/66477 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/68222 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17054 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/12236 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/10329 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/4812 "Built successfully") | [❌ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/7627 "Found 1 new failure in JSNavigationHistoryEntry.cpp") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/4831 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/8266 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/6590 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

rwlbuis commented on LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-handler-throws-expected.txt:
```diff
@@ -1,3 +1,6 @@
+CONSOLE MESSAGE: Unhandled Promise Rejection: TypeError: a message
+
+Harness Error (FAIL), message = Unhandled rejection: a message
```
This happens with quite some tests and is kind of unfortunate. For this kind of test where the handler just throws an exception we did not find a good solution. Having said that, I suppose it would still count as pass on wpt.fyi test score.

rwlbuis reviewed (approved):
LGTM.

cdumez committed:
[Navigation] Navigation.navigate() should trigger the 'beforeunload' event synchronously
https://bugs.webkit.org/show_bug.cgi?id=282593
rdar://139628818

Reviewed by Rob Buis.

Navigation.navigate() should trigger the 'beforeunload' event synchronously and quite
a few WPT tests were relying on this.

Fixing this enabled some more tests which exposed some pre-existing bugs that I'm
fixing in the same PR so it doesn't look like regressions.

In particular, we dispatch events at NavigationHistoryEntry objects but the class
doesn't subclass ActiveDOMObject to keep the JS wrapper alive. We were also failing
to grab the JSLock in one place, which was causing assertions.

* LayoutTests/TestExpectations:
Unskip some tests that are now passing.

* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigate-event/intercept-handler-throws-expected.txt:
The output looks a bit worse but I don't think this is a regression. Rather, this looks like a timing change because
the navigation now happens sooner. I will investigate separately how to make this test fully pass.

* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/back-beforeunload-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/forward-beforeunload-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/navigate-beforeunload-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/navigate-rejection-order-beforeunload-unserializablestate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/navigate-rejection-order-invalidurl-beforeunload-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/reload-beforeunload-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/reload-rejection-order-beforeunload-unserializablestate-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/navigation-api/navigation-methods/return-value/traverseTo-beforeunload-expected.txt:
Rebaseline tests that are now passing.

* Source/WebCore/loader/FrameLoadRequest.h:
(WebCore::FrameLoadRequest::isFromNavigationAPI const):
(WebCore::FrameLoadRequest::setIsFromNavigationAPI):
* Source/WebCore/loader/FrameLoader.cpp:
(WebCore::FrameLoader::stopLoading):
(WebCore::FrameLoader::loadURL):
(WebCore::FrameLoader::loadWithDocumentLoader):
(WebCore::FrameLoader::continueLoadAfterNavigationPolicy):
* Source/WebCore/loader/FrameLoader.h:
* Source/WebCore/loader/NavigationAction.h:
(WebCore::NavigationAction::isFromNavigationAPI const):
(WebCore::NavigationAction::setIsFromNavigationAPI):
* Source/WebCore/loader/NavigationScheduler.cpp:
(WebCore::ScheduledHistoryNavigationByKey::findBackForwardItemByKey):
Make the navigation policy decision synchronous when the navigation is triggered via the navigation API.
This is the only way the navigation can proceed synchronously and fire the beforeunload event right away.

* Source/WebCore/page/LocalDOMWindow.cpp:
(WebCore::LocalDOMWindow::stop):
When calling `window.stop()`, abort the pending navigation API navigation as tests are expecting this.

* Source/WebCore/page/NavigateEvent.cpp:
(WebCore::NavigateEvent::NavigateEvent):
(WebCore::NavigateEvent::create):
* Source/WebCore/page/NavigateEvent.h:
* Source/WebCore/page/NavigateEvent.idl:
Make sure the grab the JSLock before initializing m_info. Without this, I was seeing assertions
in the layout tests.

* Source/WebCore/page/Navigation.cpp:
* Source/WebCore/page/Navigation.h:
* Source/WebCore/page/NavigationHistoryEntry.cpp:
(WebCore::NavigationHistoryEntry::NavigationHistoryEntry):
(WebCore::NavigationHistoryEntry::create):
* Source/WebCore/page/NavigationHistoryEntry.h:
* Source/WebCore/page/NavigationHistoryEntry.idl:
Make NavigationHistoryEntry an ActiveDOMObject and keep its JS wrapper alive as long as the
NavigationHistoryEntry is held by the navigation object. This is important since the navigation
object may dispatch events at the NavigationHistoryEntry. Without this change, some of the tests
were asserting that the JS wrapper was missing during the event dispatching.

* Source/WebKit/WebProcess/WebCoreSupport/WebFrameLoaderClient.cpp:
(WebKit::WebFrameLoaderClient::dispatchDecidePolicyForNavigationAction):
Obey the policyDecisionMode being PolicyDecisionMode::Synchronous and use the DecidePolicyForNavigationActionSync
synchronous IPC when the navigation is triggered by the Navigation API so that the navigation can
start synchronously and yet still obey the policy decision from the client.

* Source/WebKit/WebProcess/WebCoreSupport/WebLocalFrameLoaderClient.cpp:
(WebKit::WebLocalFrameLoaderClient::dispatchDidFailProvisionalLoad):
Some of the tests abort the navigation during the provisional stage. This was hitting
an assertion in `WebLocalFrameLoaderClient::dispatchDidReachLayoutMilestone()`:
`ASSERT(!m_frame->isMainFrame() || webPage->corePage()->settings().suppressesIncrementalRendering() || m_didCompletePageTransition);`
The reason was the m_didCompletePageTransition wasn't reset to true and provisional load failure.
To address this issue, I now make sure to call `completePageTransitionIfNeeded()` in `dispatchDidFailProvisionalLoad()`.

Canonical link: https://commits.webkit.org/286919@main
